### PR TITLE
refactor: Decompose neighbor list into Pipeline of selector → masks → compactor

### PR DESCRIPTION
**Pull request by n-gao**
##  

Refactor neighbor list internals into a composable Pipeline

Replaces the monolithic `basic_neighborlist` engine with an explicit three-phase `Pipeline[D]`: a `CandidateSelector` produces a `CandidateBatch`, a sequence of `Mask` callables filters it, and a `Compactor` emits the final `Edges[D]`.

### New public types

- **`Pipeline[D]`** — the runner that wires selector → masks → compactor; all concrete neighbor list classes now build and invoke one internally.
- **`CandidateBatch[D]`** — NamedTuple carrying `Edges[D]` plus an `is_minimum_image` flag between pipeline phases; auto-registered as a JAX PyTree.
- **`PipelineContext`** — read-only struct (lh/rh in fractional coords, systems, optional rh remap) shared by every mask and compactor.
- **`CandidateSelector[D]`**, **`Mask`**, **`Compactor[D]`** — protocols defining each phase's contract.

### New selector classes

Each concrete neighbor list now exposes its candidate-emission logic as a standalone, reusable selector:

- `CellListSelector` — spatial-hash candidate emission + image replication.
- `DenseSelector` — per-system dense O(N²/K²) candidate emission + image replication.
- `AllDenseSelector` — all-pairs candidate emission + image replication.
- `InclusionGroupSelector` — pairs every particle with every other in the same inclusion segment (used by `all_connected_neighborlist`).
- `PrecomputedEdgesSelector` — wraps precomputed `Edges` for both refine variants; a `recompute_mic_shifts` flag controls whether shifts are reused or recomputed from current positions.

### New mask classes (`masks.py`)

Logic previously inlined in `_filter_candidates` is now split into independent, testable masks:

- `InBoundsMask` — drops candidates whose indices fall outside the valid inclusion-segment range.
- `InclusionMatchMask` — drops candidates whose lh/rh inclusion segments differ.
- `RemapDedupMask` — deduplicates the rh→lh remapped subset, keeping only one direction per pair.
- `DistanceCutoffMask` — drops candidates whose squared real-space distance exceeds the cutoff.
- `ExclusionMask` — drops minimum-image pairs sharing an exclusion segment; non-MIC periodic copies survive.

### New compactor classes (`compact.py`)

- `ReduceCompactor` — compresses survivors via `jnp.where(keep, size=k)` with a capacity assertion; mirrors each surviving edge with its reverse when `rh_index_remap` is set, restoring the symmetry that `RemapDedupMask` removed.
- `MaskOnlyCompactor` — preserves candidate count, replacing failing entries with OOB indices and zero shifts; used by `RefineMaskNeighborList`.

### `_core.py` cleanup

Removes `basic_neighborlist`, `_filter_candidates`, `_compact_edges`, `_build_segmentation_mask`, `_apply_distance_cutoff_w/wo_images`, and `_compute_distances_and_apply_cutoff`. Retains and exposes the shared primitives that selectors and masks depend on:

- `num_cells` (renamed from `_num_cells`)
- `Candidates` (renamed from `_Candidates`, now public)
- `replicate_for_images` — replicates candidates by image multiplicity, attaching shifts and the `is_minimum_image` flag.
- `make_batch_with_mic` — packs raw candidates with minimum-image shifts.
- `candidates_to_batch` — packs `(candidates, shifts, is_min)` into a `CandidateBatch[2]`.
- `real_distance_sq` — squared real-space distance used by `DistanceCutoffMask`.

### `RefineMaskNeighborList` and `RefineCutoffNeighborList`

Both refine classes now delegate to `Pipeline` via `PrecomputedEdgesSelector`. A new `_resolve_precomputed_inputs` helper handles the rh-overlay pattern: when a remap is supplied, rh rows are overlaid onto the corresponding lh slots so public lh-space edge indices remain valid while rh metadata (inclusion, exclusion, positions) is honored during filtering.

### Tests

Unit tests added for each new mask, both compactors, `remap_rh_to_lh`, `make_batch_with_mic`, `PrecomputedEdgesSelector` modes, and end-to-end `Pipeline` composition. New integration tests for `RefineMaskNeighborList` cover the MCMC patch pattern, disjoint rh, remapped rh inclusion/exclusion, and partial-overlap scenarios. New integration tests for `RefineCutoffNeighborList` verify that remapped and disjoint rh positions and exclusions are correctly evaluated after the overlay.

Part of #129.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#178](https://github.com/cusp-ai-oss/kUPS/pull/178) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#169** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#168** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#167** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/167?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#166** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#165** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#155** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#154** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#153** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/153?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#152** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/152?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#151** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#150** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/149?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#148** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#147** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/147?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#146** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/146?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#145** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#144** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/144?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#143** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/143?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#142** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/142?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#141** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#138** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#137** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/137?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#136** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/136?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#135** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#122** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/122?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#121** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#134** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/134?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#133** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/133?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#132** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/132?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#131** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/131?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#119** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/core/neighborlist/types.py
Comment (jonkhler) [inline src/kups/core/neighborlist/types.py:1]: types.py? or what is the way how we do it elsewhere?

**Review by jonkhler — CHANGES_REQUESTED** [review]

File: src/kups/core/neighborlist/types.py
Comment (n-gao) [reply src/kups/core/neighborlist/types.py:1]: I renamed it in the initial PR where we split it up

**Review by jonkhler — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Jun 2, 1:28 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138).
